components/dm/spec/grafana.go

```diff
@@ -145,6 +149,28 @@ func (i *GrafanaInstance) InitConfig(e executor.Executor, clusterName, clusterVe
 	return e.Transfer(fp, dst, false)
 }
 
+func (i *GrafanaInstance) initDashboards(e executor.Executor, spec GrafanaSpec, paths meta.DirPaths) error {
+	dashboardsDir := filepath.Join(paths.Deploy, "dashboards")
+	// To make this step idempotent, we need cleanup old dashboards first
+	if _, stderr, err := e.Execute(fmt.Sprintf("rm -f %s/*", dashboardsDir), false); err != nil {
+		return errors.Annotatef(err, "cleanup old dashboards: %s", string(stderr))
+	}
```

Comment:
I don't think it isn't idempotent even though we don't delete it. We will overwrite them in the subsequent steps.

Reply:
But how to delete a dashboard (eg. tidb.json) ?

Reply:
If two local dashboard dir have different files, the effect will depends on their order to apply. So it's obviously not idempotent